doc/dahdi_maint.8
.TH "DAHDI_MAINT" "8" "9 Sep 2011" "" ""

.SH NAME
dahdi_maint \- Sets Dahdi spans into maintenance mode, e.g.: loopback
.SH SYNOPSIS

.B dahdi_maint \-s \fInum\fB [options]
.B dahdi_maint <\-h|\-\-help> 

.SH DESCRIPTION

dahdi_maint uses the DAHDI_MAINT interface to set a Dahdi span (port
of a Dahdi adapter card) into loopback mode or similar maintenance mode.

.SH OPTIONS
.B \-s \-\-span \fInum\fR
.RS
The span number. Required.
.RE

.B \-l \-\-loopback <localhost|networkline|networkpayload|loopup|loopdown|off>
.RS
Loopback type. One of:
.IP localhost 4
loop back towards host
.IP networkline 4
network line loopback
.IP networkpayload 4
network payload loopback
.IP loopup 4
transmit loopup signal
.IP loopdown 4
transmit loopdown signal
.IP off 4
end loopback mode
.RE

.B \-i \-\-insert <fas|multi|crc|cas|prbs|bipolar>
.RS
Insert an error of a specific type
.RE

.SH EXAMPLES
Enable network line loopback on span 1:

  dahdi_maint \-s 1 \-\-loopback networkline

Disable network line loopback on span 1:

  dahdi_maint \-s 1 \-\-loopback off


.SH SEE ALSO
.PP
dahdi_tool(8), dahdi_cfg(8), asterisk(8).

.SH AUTHOR
.PP
This manual page was written by Tzafrir Cohen <tzafrir.cohen@xorcom.com>.
Permission is granted to copy, distribute and/or modify this document
under the terms of the GNU General Public License, Version 2 any later
version published by the Free Software Foundation.


ChangeLog
2012-04-18 Shaun Ruffell <sruffell@digium.com>

	* dahdi-linux 2.5.1 tagged.

2012-04-11 20:19 +0000 [r10656-10659]  Shaun Ruffell <sruffell@digium.com>

	* drivers/dahdi/xpp/xpp_dahdi.c, drivers/dahdi/xpp/xpp_dahdi.h:
	  xpp: Fix compilation when CONFIG_DAHDI_WATCHDOG is defined. Looks
	  like a hold over from when dahdi_span_ops was first implemented
	  in r8985 "dahdi: Move the callbacks in dahdi_span into its own
	  structure" [1]. [1]
	  http://svnview.digium.com/svn/dahdi?view=revision&revision=8985
	  Signed-off-by: Shaun Ruffell <sruffell@digium.com> Acked-by:
	  Tzafrir Cohen <tzafrir.cohen@xorcom.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10658 Conflicts:
	  drivers/dahdi/xpp/xpp_dahdi.h

	* drivers/dahdi/dahdi-base.c: dahdi: Fix compilation when
	  CONFIG_DAHDI_WATCHDOG is defined. From: Mike Sinkovsky
	  <msink@trikom.ru> Internal-Issue-ID: DAHLIN-288 Signed-off-by:
	  Shaun Ruffell <sruffell@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10655 Conflicts:
	  drivers/dahdi/dahdi-base.c

2012-04-11 09:11 +0000 [r10650-10653]  Tzafrir Cohen <tzafrir.cohen@xorcom.com>

	* drivers/dahdi/xpp/firmwares/FPGA_1161.201.hex: FPGA_1161.201.hex
	  rev 10545: fix reset of XR1000 Previous commit (r10650) included
	  an incorrect version. Including full message from that commit for
	  the description. rev. 10502 of the FPGA firmware for the new
	  E-Main rev. 4 fixes a potential issue when used on Xorcom XR1000
	  systems: an issue with the power supply may cause the unit to
	  reset. Note that there is no issue with previous models, with a
	  normal setup of an Astribank, or other XRx000 systems.
	  Signed-off-by: Tzafrir Cohen <tzafrir.cohen@xorcom.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10652

	* drivers/dahdi/xpp/firmwares/FPGA_1161.201.hex: FPGA_1161.201.hex
	  rev 10532: fix reset of XR1000 rev. 10502 of the FPGA firmware
	  for the new E-Main rev. 4 fixes a potential issue when used on
	  Xorcom XR1000 systems: an issue with the power supply may cause
	  the unit to reset. Note that there is no issue with previous
	  models, with a normal setup of an Astribank, or other XRx000
	  systems. Signed-off-by: Tzafrir Cohen <tzafrir.cohen@xorcom.com>
	  Origin: http://svnview.digium.com/svn/dahdi?view=rev&rev=10649

2012-04-05 20:34 +0000 [r10642-10647]  Shaun Ruffell <sruffell@digium.com>

	* drivers/dahdi/voicebus/Kbuild, drivers/dahdi/wct4xxp/Kbuild:
	  wcte12xp, wctdm24xxp, wct4xxp: Print warning about potential GPL
	  violation w/HOTPLUG_FIRMWARE=no. Print a warning message that it
	  may be a GPL violation to redistribute these binaries if the
	  firmware for the VPMOCT032/64/128/256 is compiled in.
	  Signed-off-by: Shaun Ruffell <sruffell@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10646 Conflicts:
	  drivers/dahdi/wct4xxp/Kbuild

	* drivers/dahdi/wcb4xxp/base.c: wcb4xxp: Remove asm/system.h
	  include. Not needed anymore and will break compilation on Kernel
	  versions >= 3.4 since commit 0195c00244dc2e [1] [1]
	  http://git.kernel.org/?p=linux/kernel/git/torvalds/linux.git;a=commit;h=0195c00244dc2e
	  Signed-off-by: Shaun Ruffell <sruffell@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10641

	* drivers/dahdi/dahdi_dummy.c: dahdi_dummy: Include timer.h instead
	  of time.h It appears that some kernel configurations do not
	  include timer.h in any of the include files that are included by
	  dahdi_dummy. The timer_structs are defined in timer.h and not
	  time.h, so this change is correct even though I never could find
	  a configuation myself that actually failed to compile. This has
	  negligible impact since dahdi_dummy is not compiled by default.
	  Internal-Issue-ID: DAHLIN-185 Reported-by: Steve Murphy
	  Signed-off-by: Shaun Ruffell <sruffell@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10640

2012-04-03 22:02 +0000 [r10635]  Shaun Ruffell <sruffell@digium.com>

	* drivers/dahdi/dahdi-base.c: dahdi: Fix compilation when
	  CONFIG_DAHDI_ECHOCAN_PROCESS_TX is defined. 'ec_state' was
	  renamed to 'dahdi_echocan_state' in r6529 [1] but support for
	  CONFIG_DAHDI_ECHOCAN_PROCESS_TX was first committed in r9442 [2].
	  So it appears that I never compiled tested this exact commit when
	  it went in for the 2.5.0 release. [1]
	  http://svnview.digium.com/svn/dahdi?view=revision&revision=6529
	  [2]
	  http://svnview.digium.com/svn/dahdi?view=revision&revision=9442
	  Reported-by: Pavel Selivanov Internal-Issue-ID: DAHLIN-279
	  Patches: ec.patch uploaded by Pavel Selivanov (License #5420) [
	  edited the patch slightly for minor formatting ] Signed-off-by:
	  Shaun Ruffell <sruffell@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10633

2012-04-02 14:05 +0000 [r10619]  Shaun Ruffell <sruffell@digium.com>

	* drivers/dahdi/voicebus/vpmoct.c, drivers/dahdi/firmware/Makefile,
	  drivers/dahdi/voicebus/Kbuild: wctdm24xxp, wcte12xp: Allow
	  VPMOCT032 firmware to be compiled into driver. Enables the driver
	  to update firmware on systems that do not have the firmware
	  loader configured / enabled (Linux config option
	  CONFIG_FW_LOADER). Compiling the firmware into the driver
	  increase the memory footprint by around ~440K. Internal-Issue-ID:
	  DAHDI-963 Reported-and-Tested-by: Guenther Kelleter
	  Signed-off-by: Shaun Ruffell <sruffell@digium.com> Origin:
	  http://svnview.digium.com/svn/dahdi?view=rev&rev=10618

2012-03-29 15:29 +0000 [r10615]  Shaun Ruffell <sruffell@digium.com>

	* drivers/dahdi/wctdm24xxp/base.c: wctdm24xxp: Remove forward
	  declaration of inline for GCC 3.4.4 GCC 3.4.4 does not allow
	  forward declaration of inline functions. Internal-Issue-ID:
	  DAHLIN-286 Reported-by: Guenther Kelleter Patches:
	  wctdm24xxp-inline.patch uploaded by Guenther Kelleter (License
	  #6372) Signed-off-by: Shaun Ruffell <sruffell@digium.com> Origin:

dahdi_maint.c
/*
 * Performance and Maintenance utility
 *
 * Written by Russ Meyerriecks <rmeyerriecks@digium.com>
 *
 * Copyright (C) 2009-2010 Digium, Inc.
 *
 * All rights reserved.
 *
 */

/*
 * See http://www.asterisk.org for more information about
 * the Asterisk project. Please do not directly contact
 * any of the maintainers of this project for assistance;
 * the project provides a web site, mailing lists and IRC
 * channels for your use.
 *
 * This program is free software, distributed under the terms of
 * the GNU General Public License Version 2 as published by the
 * Free Software Foundation. See the LICENSE file included with
 * this program for more details.
 */

#include <stdio.h>
#include <getopt.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <sys/ioctl.h>
#include <fcntl.h>
#include <stdlib.h>

#include <dahdi/user.h>
#include "dahdi_tools_version.h"

#define DAHDI_CTL "/dev/dahdi/ctl"

extern char *optarg;
extern int optind;

void display_help(char *argv0, int exitcode)
{
	char *c;
	c = strrchr(argv0, '/');
	if (!c)
		c = argv0;
	else
		c++;
	fprintf(stderr, "%s\n\n", dahdi_tools_version);
	fprintf(stderr, "Usage: %s -s <span num> <options>\n", c);
	fprintf(stderr, "Options:\n");
	fprintf(stderr, "        -h, --help		display help\n");
	fprintf(stderr, "        -s, --span <span num>	specify the span\n");
	fprintf(stderr, "        -l, --loopback <localhost|networkline|"\
						"networkpayload|loopup|"\
						"loopdown|off>\n"\
			"\t\tlocalhost - loop back towards host\n"\
			"\t\tnetworkline - network line loopback\n"\
			"\t\tnetworkpayload - network payload loopback\n"\
			"\t\tloopup - transmit loopup signal\n"\
			"\t\tloopdown - transmit loopdown signal\n"\
			"\t\toff - end loopback mode\n");
	fprintf(stderr, "        -i, --insert <fas|multi|crc|cas|prbs|bipolar>"\
			"\n\t\tinsert an error of a specific type\n");
	fprintf(stderr, "        -r, --reset		"\
			"reset the error counters\n\n");
	fprintf(stderr, "Examples: \n");
	fprintf(stderr, "Enable network line loopback\n");
	fprintf(stderr, "	dahdi_maint -s 1 --loopback networkline\n");
	fprintf(stderr, "Disable network line loopback\n");
	fprintf(stderr, "	dahdi_maint -s 1 --loopback off\n\n");

	exit(exitcode);
}

int main(int argc, char *argv[])
{
	static int ctl = -1;
	int res;

	int doloopback = 0;
	char *larg = NULL;
	int span = 1;
	int iflag = 0;
	char *iarg = NULL;
	int gflag = 0;
	int c;
	int rflag = 0;

	struct dahdi_maintinfo m;
	struct dahdi_spaninfo s;

	static struct option long_options[] = {
		{"help",	no_argument,	   0, 'h'},
		{"loopback",	required_argument, 0, 'l'},
		{"span",	required_argument, 0, 's'},
		{"insert",	required_argument, 0, 'i'},
		{"reset",	no_argument, 	   0, 'r'},
		{0, 0, 0, 0}
	};
	int option_index = 0;

	if (argc < 2) { /* no options */
		display_help(argv[0], 1);
	}

	while ((c = getopt_long(argc, argv, "hj:l:p:s:i:g:r",
				long_options, &option_index)) != -1) {
			switch (c) {
			case 'h':
				display_help(argv[0], 0);
				break;
			case 'l': /* loopback */
				larg = optarg;
				doloopback = 1;
				break;
			case 's': /* specify a span */
				span = atoi(optarg);
				break;
			case 'i': /* insert an error */
				iarg = optarg;
				iflag = 1;
				break;
			case 'g': /* generate psuedo random sequence */
				gflag = 1;
				break;
			case 'r': /* reset the error counters */
				rflag = 1;
				break;
			}
	}

	ctl = open(DAHDI_CTL, O_RDWR);
	if (ctl < 0) {
		fprintf(stderr, "Unable to open %s\n", DAHDI_CTL);
		return -1;
	}

	if (!(doloopback || iflag || gflag || rflag)) {
		s.spanno = span;
		res = ioctl(ctl, DAHDI_SPANSTAT, &s);
		if (res || ((__u32)-1 == s.fecount))
			printf("Error counters not supported by the driver"\
					" for this span\n");
		printf("Span %d:\n", span);
		printf(">Framing Errors : %d:\n", s.fecount);
		printf(">CRC Errors : %d:\n", s.crc4count);
		printf(">Code Violations : %d:\n", s.cvcount);
		printf(">E-bit Count : %d:\n", s.ebitcount);
		printf(">General Errored Seconds : %d:\n", s.errsec);

		return 0;
	}

	m.spanno = span;

	if (doloopback) {
		if (!strcasecmp(larg, "localhost")) {
			printf("Span %d: local host loopback ON\n", span);
			m.command = DAHDI_MAINT_LOCALLOOP;
		} else if (!strcasecmp(larg, "networkline")) {
			printf("Span %d: network line loopback ON\n", span);
			m.command = DAHDI_MAINT_NETWORKLINELOOP;
		} else if (!strcasecmp(larg, "networkpayload")) {
			printf("Span %d: network payload loopback ON\n", span);
			m.command = DAHDI_MAINT_NETWORKPAYLOADLOOP;
		} else if (!strcasecmp(larg, "loopup")) {
			printf("Span %d: transmitting loopup signal\n", span);
			m.command = DAHDI_MAINT_LOOPUP;
		} else if (!strcasecmp(larg, "loopdown")) {
			printf("Span %d: transmitting loopdown signal\n", span);
			m.command = DAHDI_MAINT_LOOPDOWN;
		} else if (!strcasecmp(larg, "off")) {
			printf("Span %d: loopback OFF\n", span);
			m.command = DAHDI_MAINT_NONE;
		} else {
			display_help(argv[0], 1);
		}

		res = ioctl(ctl, DAHDI_MAINT, &m);
		if (res) {
			printf("This type of looping not supported by the"\
					" driver for this span\n");
			return 1;
		}

		/* Leave the loopup/loopdown signal on the line for
		 * five seconds according to AT&T TR 54016
		 */
		if ((m.command == DAHDI_MAINT_LOOPUP) ||
		   (m.command == DAHDI_MAINT_LOOPDOWN)) {
			sleep(5);
			m.command = DAHDI_MAINT_NONE;
			ioctl(ctl, DAHDI_MAINT, &m);
		}
	}

	if (iflag) {
		if (!strcasecmp(iarg, "fas")) {
			m.command = DAHDI_MAINT_FAS_DEFECT;
			printf("Inserting a single FAS defect\n");
		} else if (!strcasecmp(iarg, "multi")) {
			m.command = DAHDI_MAINT_MULTI_DEFECT;
			printf("Inserting a single multiframe defect\n");
		} else if (!strcasecmp(iarg, "crc")) {
			m.command = DAHDI_MAINT_CRC_DEFECT;
			printf("Inserting a single CRC defect\n");
		} else if (!strcasecmp(iarg, "cas")) {
			m.command = DAHDI_MAINT_CAS_DEFECT;
			printf("Inserting a single CAS defect\n");
		} else if (!strcasecmp(iarg, "prbs")) {
			m.command = DAHDI_MAINT_PRBS_DEFECT;
			printf("Inserting a single PRBS defect\n");
		} else if (!strcasecmp(iarg, "bipolar")) {
			m.command = DAHDI_MAINT_BIPOLAR_DEFECT;
			printf("Inserting a single bipolar defect\n");
#ifdef DAHDI_MAINT_ALARM_SIM
		} else if (!strcasecmp(iarg, "sim")) {
			m.command = DAHDI_MAINT_ALARM_SIM;
			printf("Incrementing alarm simulator\n");
#endif
		} else {
			display_help(argv[0], 1);
		}
		res = ioctl(ctl, DAHDI_MAINT, &m);
		if (res)
			printf("This type of error injection is not supported"\
					" by the driver for this span\n");
	}

	if (gflag) {
		printf("Enabled the Pseudo-Random Binary Sequence Generation"\
			" and Monitor\n");
		m.command = DAHDI_MAINT_PRBS;
		res = ioctl(ctl, DAHDI_MAINT, &m);
		if (res) {
			printf("Pseudo-random binary sequence generation is"\
				" not supported by the driver for this span\n");
		}
	}

	if (rflag) {
		printf("Resetting error counters for span %d\n", span);
		m.command = DAHDI_RESET_COUNTERS;
		res = ioctl(ctl, DAHDI_MAINT, &m);
		if (res) {
			printf("Resetting error counters is not supported by"\
					" the driver for this span\n");
		}
	}

	return 0;
}
